src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/ActivatorUtilitiesTests.cs

```diff
@@ -0,0 +1,289 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using Xunit;
+using Moq;
+
+namespace Microsoft.Extensions.DependencyInjection.Tests
+{
+    public class ActivatorUtilitiesTests
+    {
+        [Fact]
+        public void CreateInstance_ClassWithABCS_UsesTheLongestAvailableConstructor()
+        {
+            var services = new ServiceCollection();
+            services.AddScoped<B>();
+            services.AddScoped<S>();
+            using var provider = services.BuildServiceProvider();
+            var a = new A();
+            var c = new C();
+
+            var instance = ActivatorUtilities.CreateInstance<ClassWithABCS>(provider, a, c);
+
+            Assert.NotNull(instance.B);
+            Assert.NotNull(instance.S);
+            Assert.Same(a, instance.A);
+            Assert.Same(c, instance.C);
+        }
+
+        [Fact]
+        public void CreateInstance_OneCtor_IsRegistered_CreatesInstanceSuccessfully()
+        {
+            var services = new ServiceCollection();
+            services.AddScoped<A>();
+            using var provider = services.BuildServiceProvider();
+
+            var instance = ActivatorUtilities.CreateInstance<ClassWithA>(provider);
+            Assert.NotNull(instance.A);
+        }
+
+        [Theory]
+        [InlineData(false)]
+        [InlineData(true)]
+        public void CreateInstance_OneCtorRegistered_IncorrectIServiceProviderIsService_CreatesInstanceSuccessfully(bool isService)
+        {
+            var mockedService = new Mock<IServiceProviderIsService>();
+            mockedService.Setup(r => r.IsService(It.IsAny<Type>()))
+                .Returns(isService);
+            var services = new ServiceCollection();
+            services.AddScoped<IServiceProviderIsService>(p => mockedService.Object);
```

Comment:
Does this actually work? Looking at the DI code, the `IServiceProviderIsService` service gets added after all the provided serviceDescriptors. Which I think means the built-in DI container's implementation will always win.

[https://github.com/dotnet/runtime/blob/3435b39194137676d9b2da12c8ebd21d69cd06b9/src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceProvider.cs#L48-L53](https://github.com/dotnet/runtime/blob/3435b39194137676d9b2da12c8ebd21d69cd06b9/src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceProvider.cs#L48-L53)